File: Sean的leetcode/FB/Implement strStr().py
###
# KMP
# Time Complexity: O(n)
# Space Complexity: O(m)
# Ref: http://www.geeksforgeeks.org/searching-for-patterns-set-2-kmp-algorithm/
###
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if not needle:
            return 0
        lps = self.kmp(needle)
        i = 0
        j = 0
        while i < len(haystack):
            if haystack[i] == needle[j]:
                i += 1
                j += 1
                if j == len(needle):
                    return i - j
            elif j:
                j = lps[j-1]
            else:
                j = 0
                i += 1
        return -1
    def kmp(self, needle):
        lps = [0] * len(needle)
        j = 0
        i = 1
        while i < len(needle):
            if needle[j] == needle[i]:
                j += 1
                lps[i] = j
                i += 1
            elif j:
                j = lps[j-1]
            else:
                j = 0
                i += 1
        return lps       
###
# Time Complexity: O(n*m)
###
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        if not needle:
            return 0
        for i in xrange(len(haystack) - len(needle) + 1):
            if haystack[i:i+len(needle)] == needle:
                return i
        return -1
        

File: MyLeetCode/Leetcode/Implement strStr().cpp
#if 1
class Solution {
public:
    int strStr(string haystack, string needle) {
        int len1 = haystack.size();
        int len2 = needle.size();
        int count = 0;

        if (haystack == needle || needle.size() == 0)
            return 0;
        for (int i = 0, j = 0; i < len1;)
        {
            if (haystack[i] == needle[j])
            {
                i++, j++, count++;
            }
            else
            {
                i = i - count + 1;
                j = count = 0;
            }

            if (count == len2)
            {
                return i - count;
            }
        }
        return -1;
    }
};
#else
class Solution {
public:
    int strStr(string haystack, string needle) {
        int len1 = haystack.size();
        int len2 = needle.size();

        for (int i = 0; i < len1 - (len2 - 1); i++)
        {
            string sub = haystack.substr(i, len2);
            if (sub == needle)
                return i;
        }
        return -1;
    }
};
#endif